python/ray/experimental/data/impl/sort.py

```diff
@@ -0,0 +1,116 @@
+"""
+TODO: document the algorithm
+"""
+from typing import List, Any, Callable, TypeVar, Tuple, Union
+
+import numpy as np
+import pyarrow as pa
+import ray
+from ray.experimental.data.block import Block, BlockAccessor
+from ray.experimental.data.impl.block_list import BlockList
+from ray.experimental.data.impl.progress_bar import ProgressBar
+
+T = TypeVar("T")
+
+# Data can be sorted by value (None), a column (str), multiple columns
+# (List[str]) or a custom transform function (lambda).
+SortKeyT = Union[None, str, List[str], Callable[[T], Any]]
+
+
+def sample_boundaries(blocks: BlockList[T], key: SortKeyT,
+                      num_reducers: int) -> List[T]:
+    """return len(boundaries) == num_reducers - 1"""
+    n_samples = int(num_reducers * 10 / len(blocks))
+
+    def _sample_simple_block(items: List[T]) -> np.ndarray:
+        k = min(n_samples, len(items))
+        samples = np.random.choice(items, k, replace=False)
+        vf = np.vectorize(lambda x: key(x) if key else x)
+        return vf(samples)
+
+    # TODO: type this
+    def _sample_arrow_block(table: pa.Table):
+        k = min(n_samples, table.num_rows)
+        indices = np.random.choice(table.num_rows, k, replace=False)
+        # TODO: handle multi column
```

Comment:
For pyarrow.Table, shall we raise NotImplementedError for multi-column / key function for now?

For the list block, we can raise NotImplementedError for the column key types.